keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,345 @@
+
+## Non-Goals
+
+- Updating scheduler, pod spec is out of scope at this point.
+
+- This proposal only focus on Linux based systems.
+
+## User Stories
+
+### Story 1 : Networking Acceleration using DPDK
+
+- The system such as real-time trading system and 5G system, which requires networking acceleration, uses DPDK to guarantee low latency of packet processing. DPDK(Data Plane Development Kit) is set of libraries to accelerate packet processing on userspace. DPDK requires dedicated resources(such as exclusive cpu usage, hugepages and DPDK compatile Network Interface Card) and alignment of resources to sigle NUMA node to accelerate its performance. There should be a way to guarantee reserve Memory and Hugepages with other computing resources on desinated Socket for containerized Network Function which uses DPDK.
+
+### Story 2 : Database
+
+- Databases(Oracle, PostgreSQL, MySQL, MongoDB, SAP) consumes massive Memory and Hugepages, in order to reduce memory access latency and improve performance dramatically resources(CPU, Memory, Hugepages, I/O Devices) should be aligned.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+Memory Manager is a new component of Kubelet, it guarantees NUMA aware memory usage for the Pod which has Guarantedd QoS, here memory includes hugepages.
+
+To guarantee NUMA aware memory usage, when Guaranteed QoS Pod admission is requested, Memory Manager caculates NUMA node affinity to figure out appropriate NUMA node where enough memory and hugepages exist for containers in the Pod. Then, when Pod deployed, it enfoces containers in the Pod to consume memory and hugepages from designated NUMA node which Topology Manager selected.
+
+NOTE:
+- Memory Manager isolates Memory and Hugepages to single NUMA node to prevent remote(Inter-NUMA) memory access. For this, Memory Manager enforce cgroups cpuset subsystem using CRI API. For more details, see [CPUSET Enforcement](#cpuset-enforcement).
+
+- Memory Manager guarantees that the container(in Guanteed QoS Pod) consumes requested amount of memory(which is isolated to a NUMA node) through linux OOM system. For more details, see [OOM Score Enforcement](#oom-score-enforcement)
+
+More details of this component are listed below.
+
+#### The Concept of Safety Zone
+
+Memory Manager has the concept of safety zone to manage memory and hugepages at the NUMA node level. Basically, Kubernetes support reserving memory at the node level through Node Allocatable feature. It provides options such as `system-reserved` and `kube-reserved` to reserve CPUs, Memory, Storage for system services and kubernetes system components. Fortunately, it is possible to reserve CPUs with consideration of NUMA because CPUs are reserved using their logical core number. On the other hand, it is impossible to reserve memory with consideration of NUMA, the only thing we can do is reserving memory at the node level.
```

Comment:
Why is it "impossible to reserve memory with consideration of NUMA"?  Do you just mean that it is impossible given the existing flags for kubelet but would be possible if we changed those flags?

Reply:
@cbf123 
` Do you just mean that it is impossible given the existing flags for kubelet but would be possible if we changed those flags?`
=> Nope,  we cannot make it with just changing the flags.
When a process attempts memory allocation using `malloc` families, Linux kernel doesn't allocate physical pages. 
The kernel just `commits` memory to make a process think it got pages.
Then the kernel allocates real pages when a process attempts writing data to memory.

So, it means we don't know that when the Linux process take pages in runtime, also which NUMA node the process will take pages from in multi-socket machine.

To guarantee a container take pages from a NUMA node, we have to do:
1) something should know the amount of allocatable memory for each NUMA node.
2) schedule a container to specific NUMA node where it has adequate memory to serve a container.
3) then pin memory to specific NUMA node using cpuset.mems

below things(safety zone, allocatable zone) are proposed for `1)`.

Reply:
While I agree with everything you've said, I don't see how that prevents you from reserving memory in a NUMA-aware way.  It does add the requirement that in order to do it every Kubernetes container would need to be affined to a single NUMA node.